swmm-toolkit/src/swmm/toolkit/solver.i

```diff
@@ -24,234 +24,101 @@
 #define SWIG_FILE_WITH_INIT
 
 #include "swmm5.h"
-#include "toolkitAPI.h"
-#include "toolkitAPI_enums.h"
+#include "toolkit.h"
+#include "toolkit_enums.h"
+#include "toolkit_structs.h"
 %}
 
 
+%include "stats_typemaps.i";
+
+
 /* TYPEMAP FOR IGNORING INT ERROR CODE RETURN VALUE */
 %typemap(out) int {
     $result = Py_None;
     Py_INCREF($result);
 }
 
+
 %apply int *OUTPUT {
+    int *index,
+    int *value,
+    int *count, 
+    int *node1, 
+    int *node2,
+    int *out_index,
+    int *condition,
     int *year,
     int *month,
     int *day,
     int *hour,
     int *minute,
     int *second
-};
+}
+
 
 %apply double *OUTPUT {
-    double *value
-};
+    double *elapsedTime,
+    double *value,
+    double *result
+}
+
+
+%apply float *OUTPUT {
+    float *runoffErr,
+    float *flowErr, 
+    float *qualErr
+}
+
 
 %apply signed char *OUTPUT {
     signed char *value
-};
+}
+
 
 %apply char *OUTPUT {
     char *condition
-};
+}
+
 
 %cstring_output_allocate(char **OUTCHAR, swmm_freeMemory(*$1));
 
 %apply char **OUTCHAR {
-    char **errorMsg,
     char **id,
     char **major, 
     char **minor, 
     char **patch
-};
+}
+
 
 /* TYPEMAPS FOR MEMORY MANAGEMNET OF DOUBLE ARRAYS */
-%typemap(in, numinputs=0)double** in_out_array (double* temp){
-   $1 = &temp;
+%typemap(in, numinputs=0) double **double_out (double *temp) {
+    $1=&temp;
+}
+%typemap(in, numinputs=0) int *int_dim (int temp) {
+    $1=&temp;
 }
-%typemap(argout) (double** in_out_array) {
-    if (*$1)
-    {
-      int length = sizeof(*$1)/sizeof(double) + 1;
-      PyObject *o = PyList_New(length);
-      double* temp = *$1;
-      for(int i=0; i<length; i++) {
-        PyList_SetItem(o, i, PyFloat_FromDouble(temp[i]));
+%typemap(argout) (double **double_out, int *int_dim) {
+    if (*$1) {
+      PyObject *o = PyList_New(*$2);
+      for(int i=0; i<*$2; i++) {
+        PyList_SetItem(o, i, PyFloat_FromDouble(temp$argnum[i]));
```

Comment:
What does temp$argnum mean?

Reply:
@jennwuu So this typemap creates a temp variable for one of the function arguments. When it creates the temp variable it gets appended with the argument number so it has a unique name. You can use `$argnum` when you generate the wrapper to reference the temp variable you created in a consistent and generic way. So if I apply this typemap to argument 4 then the `temp4` variable would get created and referenced in the typemap. If you haven't already you really should study the wrapper code that SWIG generates. It makes more sense when you actually see what SWIG is doing.